# HG changeset patch
# User Robert Relyea <rrelyea@redhat.com>
# Date 1761684967 25200
#      Tue Oct 28 13:56:07 2025 -0700
# Branch RHEL10
# Node ID 41c115659ed4c65119d97e4be6bdc3030c9a4586
# Parent  a2ec761c754f728b669dc32ea94e260362b43b5b
nss-3.112-mlkem-fips-update.patch

diff --git a/lib/freebl/fipsfreebl.c b/lib/freebl/fipsfreebl.c
--- a/lib/freebl/fipsfreebl.c
+++ b/lib/freebl/fipsfreebl.c
@@ -1797,29 +1797,34 @@ loser:
 
 static SECStatus
 freebl_ML_KEM_Test(KyberParams param_set,
                    const unsigned char *seed, size_t seed_len,
                    const unsigned char *enc_seed, size_t enc_seed_len,
                    const unsigned char *pub_key, size_t pub_key_len,
                    const unsigned char *priv_key, size_t priv_key_len,
                    const unsigned char *cipher_text, size_t cipher_text_len,
-                   const unsigned char *key, size_t key_len)
+                   const unsigned char *key, size_t key_len,
+                   const unsigned char *rej_key, size_t rej_key_len)
 {
     SECStatus rv;
     unsigned char cipher_text_buf[MAX_ML_KEM_CIPHER_LENGTH];
+    unsigned char zero_text_buf[MAX_ML_KEM_CIPHER_LENGTH] = { 0 };
     unsigned char priv_key_buf[MAX_ML_KEM_PRIVATE_KEY_LENGTH];
     unsigned char pub_key_buf[MAX_ML_KEM_PUBLIC_KEY_LENGTH];
     unsigned char key_buf[KYBER_SHARED_SECRET_BYTES];
     unsigned char key2_buf[KYBER_SHARED_SECRET_BYTES];
+    unsigned char key3_buf[KYBER_SHARED_SECRET_BYTES];
     SECItem ct_item = { siBuffer, cipher_text_buf, cipher_text_len };
+    SECItem zt_item = { siBuffer, zero_text_buf, cipher_text_len };
     SECItem priv_key_item = { siBuffer, priv_key_buf, priv_key_len };
     SECItem pub_key_item = { siBuffer, pub_key_buf, pub_key_len };
     SECItem key_item = { siBuffer, key_buf, sizeof(key_buf) };
     SECItem key2_item = { siBuffer, key2_buf, sizeof(key2_buf) };
+    SECItem key3_item = { siBuffer, key3_buf, sizeof(key3_buf) };
     SECItem seed_item = { siBuffer, (unsigned char *)seed, seed_len };
     SECItem eseed_item = { siBuffer, (unsigned char *)enc_seed, enc_seed_len };
 
     PORT_Assert(pub_key_len <= sizeof(pub_key_buf));
     PORT_Assert(priv_key_len <= sizeof(priv_key_buf));
     PORT_Assert(cipher_text_len <= sizeof(cipher_text_buf));
 
     rv = Kyber_NewKey(param_set, &seed_item, &priv_key_item, &pub_key_item);
@@ -1828,43 +1833,53 @@ freebl_ML_KEM_Test(KyberParams param_set
         return rv;
     }
 
     if ((priv_key_item.len != priv_key_len) ||
         (pub_key_item.len != pub_key_len) ||
         (PORT_Memcmp(priv_key_item.data, priv_key, priv_key_len) != 0) ||
         (PORT_Memcmp(pub_key_item.data, pub_key, pub_key_len) != 0)) {
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
-        return rv;
+        return SECFailure;
     }
 
     rv = Kyber_Encapsulate(param_set,  &eseed_item, &pub_key_item,
                            &ct_item, &key_item);
     if (rv != SECSuccess) {
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
         return rv;
     }
 
     if ((ct_item.len != cipher_text_len) ||
         (key_item.len != key_len) ||
         (PORT_Memcmp(ct_item.data, cipher_text, cipher_text_len) != 0) ||
         (PORT_Memcmp(key_item.data, key, key_len) != 0)) {
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
-        return rv;
+        return SECFailure;
     }
 
     rv = Kyber_Decapsulate(param_set,  &priv_key_item, &ct_item, &key2_item);
     if (rv != SECSuccess) {
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
         return rv;
     }
     if (SECITEM_CompareItem(&key2_item, &key_item) != 0) {
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
+        return SECFailure;
+    }
+    rv = Kyber_Decapsulate(param_set,  &priv_key_item, &zt_item, &key3_item);
+    if (rv != SECSuccess) {
+        PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
         return rv;
     }
+    if ((key3_item.len != rej_key_len) ||
+        (PORT_Memcmp(key3_item.data, rej_key, rej_key_len) != 0)) {
+        PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
+        return SECFailure;
+    }
     return SECSuccess;
 }
 
 static SECStatus
 freebl_fips_ML_KEM_PowerUpSelfTest()
 {
     SECStatus rv;
 
@@ -1881,27 +1896,29 @@ freebl_fips_ML_KEM_PowerUpSelfTest()
     PORT_Assert(sizeof (ml_kem1024_key) == KYBER_SHARED_SECRET_BYTES);
 
     rv = freebl_ML_KEM_Test(params_ml_kem768_test_mode,
                             ml_kem_key_seed, KYBER_KEYPAIR_COIN_BYTES,
                             ml_kem_enc_seed, KYBER_SHARED_SECRET_BYTES,
                             ml_kem768_pub_key, KYBER768_PUBLIC_KEY_BYTES,
                             ml_kem768_priv_key, KYBER768_PRIVATE_KEY_BYTES,
                             ml_kem768_cipher_text, KYBER768_CIPHERTEXT_BYTES,
-                            ml_kem768_key, KYBER_SHARED_SECRET_BYTES);
+                            ml_kem768_key, KYBER_SHARED_SECRET_BYTES,
+                            ml_kem768_rej_key, KYBER_SHARED_SECRET_BYTES);
     if (rv != SECSuccess) {
         return SECFailure;
     }
     rv = freebl_ML_KEM_Test(params_ml_kem1024_test_mode,
                             ml_kem_key_seed, KYBER_KEYPAIR_COIN_BYTES,
                             ml_kem_enc_seed, KYBER_SHARED_SECRET_BYTES,
                             ml_kem1024_pub_key, MLKEM1024_PUBLIC_KEY_BYTES,
                             ml_kem1024_priv_key, MLKEM1024_PRIVATE_KEY_BYTES,
                             ml_kem1024_cipher_text, MLKEM1024_CIPHERTEXT_BYTES,
-                            ml_kem1024_key, KYBER_SHARED_SECRET_BYTES);
+                            ml_kem1024_key, KYBER_SHARED_SECRET_BYTES,
+                            ml_kem1024_rej_key, KYBER_SHARED_SECRET_BYTES);
     if (rv != SECSuccess) {
         return SECFailure;
     }
 
     return SECSuccess;
 }
 
 #ifdef NSS_ENABLE_ML_DSA
diff --git a/lib/freebl/kyber.c b/lib/freebl/kyber.c
--- a/lib/freebl/kyber.c
+++ b/lib/freebl/kyber.c
@@ -4,16 +4,17 @@
 
 #ifdef FREEBL_NO_DEPEND
 #include "stubs.h"
 #endif
 
 #include <stdbool.h>
 
 #include "blapi.h"
+#include "blapii.h"
 #include "secerr.h"
 #include "secitem.h"
 
 #include "kyber-pqcrystals-ref.h"
 #include "kyber.h"
 #include "verified/internal/libcrux_core.h"
 #include "verified/libcrux_mlkem768_portable.h"
 #include "verified/libcrux_mlkem768.h"
@@ -198,29 +199,39 @@ Kyber_NewKey(KyberParams params, const S
         }
         coins = randbuf;
     }
     NSS_CLASSIFY(coins, KYBER_KEYPAIR_COIN_BYTES);
     if (params == params_ml_kem768 || params == params_ml_kem768_test_mode) {
         libcrux_ml_kem_mlkem768_MlKem768KeyPair keys = libcrux_ml_kem_mlkem768_portable_generate_key_pair(coins);
         memcpy(pubkey->data, keys.pk.value, KYBER768_PUBLIC_KEY_BYTES);
         memcpy(privkey->data, keys.sk.value, KYBER768_PRIVATE_KEY_BYTES);
+        PORT_SafeZero(&keys, sizeof(keys));
     } else if (params == params_ml_kem1024 || params == params_ml_kem1024_test_mode) {
         libcrux_ml_kem_mlkem1024_MlKem1024KeyPair keys = libcrux_ml_kem_mlkem1024_portable_generate_key_pair(coins);
         memcpy(pubkey->data, keys.pk.value, MLKEM1024_PUBLIC_KEY_BYTES);
         memcpy(privkey->data, keys.sk.value, MLKEM1024_PRIVATE_KEY_BYTES);
+        PORT_SafeZero(&keys, sizeof(keys));
 #ifndef NSS_DISABLE_KYBER
     } else if (params == params_kyber768_round3 || params == params_kyber768_round3_test_mode) {
         pqcrystals_kyber768_ref_keypair_derand(pubkey->data, privkey->data, coins);
 #endif
     } else {
         /* unreachable */
+        if (coins == randbuf) {
+            PORT_SafeZero(randbuf, sizeof(randbuf));
+        }
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
         return SECFailure;
     }
+    /* manual max stack count 68000 */
+    BLAPI_CLEAR_STACK(70000);
+    if (coins == randbuf) {
+        PORT_SafeZero(randbuf, sizeof(randbuf));
+    }
     NSS_DECLASSIFY(pubkey->data, pubkey->len);
     return SECSuccess;
 }
 
 SECStatus
 Kyber_Encapsulate(KyberParams params, const SECItem *enc_seed, const SECItem *pubkey, SECItem *ciphertext, SECItem *secret)
 {
     if (!valid_params(params)) {
@@ -239,103 +250,133 @@ Kyber_Encapsulate(KyberParams params, co
         coins = enc_seed->data;
     } else {
         if (RNG_GenerateGlobalRandomBytes(randbuf, sizeof randbuf) != SECSuccess) {
             PORT_SetError(SEC_ERROR_NEED_RANDOM);
             return SECFailure;
         }
         coins = randbuf;
     }
+    SECStatus rv = SECSuccess;
     NSS_CLASSIFY(coins, KYBER_ENC_COIN_BYTES);
     if (params == params_ml_kem768 || params == params_ml_kem768_test_mode) {
         /* shouldn't this just use the typedef im libcrux_mlkem768.h? */
         libcrux_ml_kem_types_MlKemPublicKey_15 pk_value;
         memcpy(pk_value.value, pubkey->data, KYBER768_PUBLIC_KEY_BYTES);
 
         bool valid_pk = libcrux_ml_kem_mlkem768_portable_validate_public_key(&pk_value);
         if (!valid_pk) {
+            PORT_SafeZero(&pk_value, sizeof(pk_value));
             PORT_SetError(SEC_ERROR_INVALID_ARGS);
-            return SECFailure;
+            rv = SECFailure;
+            goto loser;
         }
 
         tuple_3c encap = libcrux_ml_kem_mlkem768_portable_encapsulate(&pk_value, coins);
         memcpy(ciphertext->data, encap.fst.value, KYBER768_CIPHERTEXT_BYTES);
         memcpy(secret->data, encap.snd, KYBER_SHARED_SECRET_BYTES);
+        PORT_SafeZero(&pk_value, sizeof(pk_value));
+        PORT_SafeZero(&encap, sizeof(encap));
     } else if (params == params_ml_kem1024 || params == params_ml_kem1024_test_mode) {
         /* shouldn't this just use the typedef im libcrux_mlkem1024.h? */
         libcrux_ml_kem_types_MlKemPublicKey_1f pk_value;
         memcpy(pk_value.value, pubkey->data, MLKEM1024_PUBLIC_KEY_BYTES);
 
         bool valid_pk = libcrux_ml_kem_mlkem1024_portable_validate_public_key(&pk_value);
         if (!valid_pk) {
+            PORT_SafeZero(&pk_value, sizeof(pk_value));
             PORT_SetError(SEC_ERROR_INVALID_ARGS);
-            return SECFailure;
+            rv = SECFailure;
+            goto loser;
         }
 
         tuple_21 encap = libcrux_ml_kem_mlkem1024_portable_encapsulate(&pk_value, coins);
         memcpy(ciphertext->data, encap.fst.value, MLKEM1024_CIPHERTEXT_BYTES);
         memcpy(secret->data, encap.snd, KYBER_SHARED_SECRET_BYTES);
+        PORT_SafeZero(&pk_value, sizeof(pk_value));
+        PORT_SafeZero(&encap, sizeof(encap));
 #ifndef NSS_DISABLE_KYBER
     } else if (params == params_kyber768_round3 || params == params_kyber768_round3_test_mode) {
         pqcrystals_kyber768_ref_enc_derand(ciphertext->data, secret->data, pubkey->data, coins);
 #endif
     } else {
         /* unreachable */
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
-        return SECFailure;
+        rv = SECFailure;
+        goto loser;
     }
+loser:
+    if (coins == randbuf) {
+        PORT_SafeZero(randbuf, sizeof(randbuf));
+    }
+    /* manual max stack count 6700 */
+    BLAPI_CLEAR_STACK(10000);
 
-    return SECSuccess;
+    return rv;
 }
 
 SECStatus
 Kyber_Decapsulate(KyberParams params, const SECItem *privkey, const SECItem *ciphertext, SECItem *secret)
 {
     if (!valid_params(params)) {
         PORT_SetError(SEC_ERROR_INVALID_ALGORITHM);
         return SECFailure;
     }
 
     if (!(valid_privkey(params, privkey) && valid_ciphertext(params, ciphertext) && valid_secret(params, secret))) {
         PORT_SetError(SEC_ERROR_INVALID_ARGS);
         return SECFailure;
     }
 
+    SECStatus rv = SECSuccess;
     if (params == params_ml_kem768 || params == params_ml_kem768_test_mode) {
         libcrux_ml_kem_types_MlKemPrivateKey_55 private_key;
         memcpy(private_key.value, privkey->data, KYBER768_PRIVATE_KEY_BYTES);
 
         libcrux_ml_kem_mlkem768_MlKem768Ciphertext cipher_text;
         memcpy(cipher_text.value, ciphertext->data, KYBER768_CIPHERTEXT_BYTES);
 
         bool valid = libcrux_ml_kem_mlkem768_portable_validate_private_key(&private_key, &cipher_text);
         if (!valid) {
+            PORT_SafeZero(&private_key, sizeof(private_key));
+            PORT_SafeZero(&cipher_text, sizeof(cipher_text));
             PORT_SetError(SEC_ERROR_INVALID_ARGS);
-            return SECFailure;
+            rv = SECFailure;
+            goto loser;
         }
 
         libcrux_ml_kem_mlkem768_portable_decapsulate(&private_key, &cipher_text, secret->data);
+        PORT_SafeZero(&private_key, sizeof(private_key));
+        PORT_SafeZero(&cipher_text, sizeof(cipher_text));
     } else if (params == params_ml_kem1024 || params == params_ml_kem1024_test_mode) {
         libcrux_ml_kem_types_MlKemPrivateKey_95 private_key;
         memcpy(private_key.value, privkey->data, MLKEM1024_PRIVATE_KEY_BYTES);
 
         libcrux_ml_kem_mlkem1024_MlKem1024Ciphertext cipher_text;
         memcpy(cipher_text.value, ciphertext->data, MLKEM1024_CIPHERTEXT_BYTES);
 
         bool valid = libcrux_ml_kem_mlkem1024_portable_validate_private_key(&private_key, &cipher_text);
         if (!valid) {
+            PORT_SafeZero(&private_key, sizeof(private_key));
+            PORT_SafeZero(&cipher_text, sizeof(cipher_text));
             PORT_SetError(SEC_ERROR_INVALID_ARGS);
-            return SECFailure;
+            rv = SECFailure;
+            goto loser;
         }
 
         libcrux_ml_kem_mlkem1024_portable_decapsulate(&private_key, &cipher_text, secret->data);
+        PORT_SafeZero(&private_key, sizeof(private_key));
+        PORT_SafeZero(&cipher_text, sizeof(cipher_text));
 #ifndef NSS_DISABLE_KYBER
     } else if (params == params_kyber768_round3 || params == params_kyber768_round3_test_mode) {
         pqcrystals_kyber768_ref_dec(secret->data, ciphertext->data, privkey->data);
 #endif
     } else {
         // unreachable
         PORT_SetError(SEC_ERROR_LIBRARY_FAILURE);
         return SECFailure;
     }
+loser:
+    /* manual max stack count 8200 */
+    BLAPI_CLEAR_STACK(10000);
 
-    return SECSuccess;
+    return rv;
 }
diff --git a/lib/freebl/ml_kem_fips_vectors.h b/lib/freebl/ml_kem_fips_vectors.h
--- a/lib/freebl/ml_kem_fips_vectors.h
+++ b/lib/freebl/ml_kem_fips_vectors.h
@@ -415,16 +415,21 @@
         0xd6, 0x18, 0xf6, 0x5a, 0x70, 0x0b, 0x41, 0xf0, 0x17, 0x08, 0x67, 0x98,
         0x2f, 0x5a, 0x68, 0xc7, 0x5c, 0x65, 0x15, 0x0e,
     };
     static const unsigned char ml_kem768_key[] = {
         0x42, 0xf5, 0x58, 0xb0, 0xbc, 0x5d, 0x70, 0x0a, 0x91, 0x1b, 0x0f, 0xc6,
         0x7f, 0x62, 0x37, 0x6f, 0x7a, 0xee, 0x46, 0x67, 0xf1, 0x96, 0x9e, 0x03,
         0xf1, 0x8b, 0xdf, 0xdf, 0x3c, 0x59, 0xfb, 0xdc,
     };
+    static const unsigned char ml_kem768_rej_key[] = {
+        0x6b, 0x24, 0x6e, 0xb1, 0x8b, 0x55, 0x3a, 0xb4, 0xcc, 0x4a, 0x57, 0x04,
+        0x4d, 0x8d, 0x15, 0x34, 0xa8, 0x85, 0xd1, 0xd3, 0x42, 0x14, 0xc9, 0x87,
+        0x78, 0x96, 0x19, 0xcb, 0x73, 0xe6, 0x12, 0x9a,
+    };
     static const unsigned char ml_kem1024_pub_key[] = {
         0x4b, 0x94, 0xc2, 0x94, 0x50, 0x11, 0x11, 0x91, 0x82, 0x3b, 0x35, 0x14,
         0xc9, 0xac, 0x1e, 0xa3, 0xd9, 0x82, 0x5c, 0xcb, 0x86, 0x39, 0x3a, 0x2d,
         0xfb, 0x04, 0x65, 0x4f, 0xa2, 0x19, 0x2d, 0x37, 0xbf, 0xad, 0x1c, 0x49,
         0x7c, 0x65, 0x02, 0xee, 0xe5, 0xca, 0x80, 0xa7, 0x3b, 0xfc, 0xe0, 0xba,
         0xf5, 0xa5, 0x4a, 0x88, 0x58, 0x5a, 0x40, 0x13, 0x97, 0xa3, 0xd2, 0x32,
         0xf4, 0x26, 0xa7, 0xaf, 0xb0, 0x82, 0xbc, 0x21, 0xa4, 0x43, 0x17, 0x09,
         0x0e, 0xaa, 0xc7, 0x59, 0x2c, 0x2e, 0xa8, 0x8a, 0x65, 0x3c, 0x44, 0x91,
@@ -952,8 +957,13 @@
         0x3b, 0xaf, 0x03, 0xad, 0xf1, 0xab, 0x8f, 0x86, 0xea, 0x46, 0x28, 0xf8,
         0x53, 0x80, 0xee, 0x89, 0x53, 0xd4, 0xa1, 0xad,
     };
     static const unsigned char ml_kem1024_key[] = {
         0x73, 0x8a, 0x88, 0xcd, 0x9a, 0x25, 0x86, 0x1e, 0x9a, 0x37, 0xc6, 0x5a,
         0x67, 0xec, 0x1b, 0x39, 0xf0, 0x69, 0xae, 0xd9, 0x64, 0x6e, 0xde, 0x89,
         0x41, 0x0a, 0xce, 0x47, 0xa9, 0x6a, 0x9b, 0x57,
     };
+    static const unsigned char ml_kem1024_rej_key[] = {
+        0x6a, 0x58, 0xe1, 0xc5, 0x96, 0x86, 0xc9, 0xfc, 0x32, 0x2e, 0xc2, 0x6b,
+        0x6e, 0xa0, 0xd5, 0x27, 0x22, 0x67, 0x29, 0x2c, 0xe9, 0xa4, 0x0f, 0x12,
+        0x75, 0x13, 0x96, 0xbc, 0x53, 0x22, 0x72, 0x46,
+    };
diff --git a/lib/softoken/kem.c b/lib/softoken/kem.c
--- a/lib/softoken/kem.c
+++ b/lib/softoken/kem.c
@@ -198,16 +198,18 @@ NSC_EncapsulateKey(CK_SESSION_HANDLE hSe
     SFTKAttribute *encapsulationKey = NULL;
 
     CK_RV crv;
     SFTKFreeStatus status;
     CK_ULONG paramSet = 0; /* use a generic U_LONG so we can handle
                             * different param set types based on the
                             * Mechanism value */
     KyberParams kyberParams;
+    CK_OBJECT_CLASS ckclass = CKO_SECRET_KEY;
+    CK_KEY_TYPE ckkeyType = CKK_GENERIC_SECRET;
 
     CHECK_FORK();
 
     if (!pMechanism || !phKey || !pulCiphertextLen) {
         return CKR_ARGUMENTS_BAD;
     }
 
     if (!sftk_kem_ValidateMechanism(pMechanism)) {
@@ -232,16 +234,25 @@ NSC_EncapsulateKey(CK_SESSION_HANDLE hSe
     }
     for (unsigned long int i = 0; i < ulAttributeCount; i++) {
         crv = sftk_AddAttributeType(key, sftk_attr_expand(&pTemplate[i]));
         if (crv != CKR_OK) {
             goto cleanup;
         }
     }
 
+    crv = sftk_defaultAttribute(key, CKA_CLASS, &ckclass, sizeof(ckclass));
+    if (crv != CKR_OK) {
+        goto cleanup;
+    }
+    crv = sftk_defaultAttribute(key, CKA_KEY_TYPE, &ckkeyType, sizeof(ckkeyType));
+    if (crv != CKR_OK) {
+        goto cleanup;
+    }
+
     encapsulationKeyObject = sftk_ObjectFromHandle(hPublicKey, session);
     if (encapsulationKeyObject == NULL) {
         crv = CKR_KEY_HANDLE_INVALID;
         goto cleanup;
     }
     encapsulationKey = sftk_FindAttribute(encapsulationKeyObject, CKA_VALUE);
     if (encapsulationKey == NULL) {
         crv = CKR_KEY_HANDLE_INVALID;
@@ -264,17 +275,17 @@ NSC_EncapsulateKey(CK_SESSION_HANDLE hSe
     SECItem pubKey = { siBuffer, encapsulationKey->attrib.pValue, encapsulationKey->attrib.ulValueLen };
 
     /* The length of secretBuf can be increased if we ever support other KEMs
      * by changing the define at the top of this file */
     uint8_t secretBuf[MAX_KEM_SHARED_SECRET_BYTES] = { 0 };
     SECItem secret = { siBuffer, secretBuf, sizeof secretBuf };
 
     key->isFIPS = sftk_operationIsFIPS(slot, pMechanism, CKA_ENCAPSULATE,
-                                       key, 0);
+                                       encapsulationKeyObject, 0);
     key->source = SFTK_SOURCE_KEA;
     switch (pMechanism->mechanism) {
 #ifndef NSS_DISABLE_KYBER
         case CKM_NSS_KYBER:
 #endif
         case CKM_NSS_ML_KEM:
         case CKM_ML_KEM:
             PORT_Assert(secret.len >= KYBER_SHARED_SECRET_BYTES);
@@ -347,16 +358,18 @@ NSC_DecapsulateKey(CK_SESSION_HANDLE hSe
     SFTKAttribute *decapsulationKey = NULL;
     CK_ULONG paramSet = 0; /* use a generic U_LONG so we can handle
                             * different param set types based on the
                             * Mechanism value */
 
     CK_RV crv;
     SFTKFreeStatus status;
     KyberParams kyberParams;
+    CK_OBJECT_CLASS ckclass = CKO_SECRET_KEY;
+    CK_KEY_TYPE ckkeyType = CKK_GENERIC_SECRET;
 
     CHECK_FORK();
 
     if (!pMechanism || !pCiphertext || !pTemplate || !phKey) {
         return CKR_ARGUMENTS_BAD;
     }
 
     if (!sftk_kem_ValidateMechanism(pMechanism)) {
@@ -381,16 +394,25 @@ NSC_DecapsulateKey(CK_SESSION_HANDLE hSe
     }
     for (unsigned long int i = 0; i < ulAttributeCount; i++) {
         crv = sftk_AddAttributeType(key, sftk_attr_expand(&pTemplate[i]));
         if (crv != CKR_OK) {
             goto cleanup;
         }
     }
 
+    crv = sftk_defaultAttribute(key, CKA_CLASS, &ckclass, sizeof(ckclass));
+    if (crv != CKR_OK) {
+        goto cleanup;
+    }
+    crv = sftk_defaultAttribute(key, CKA_KEY_TYPE, &ckkeyType, sizeof(ckkeyType));
+    if (crv != CKR_OK) {
+        goto cleanup;
+    }
+
     decapsulationKeyObject = sftk_ObjectFromHandle(hPrivateKey, session);
     if (decapsulationKeyObject == NULL) {
         crv = CKR_KEY_HANDLE_INVALID;
         goto cleanup;
     }
     decapsulationKey = sftk_FindAttribute(decapsulationKeyObject, CKA_VALUE);
     if (decapsulationKey == NULL) {
         crv = CKR_KEY_HANDLE_INVALID;
@@ -412,17 +434,17 @@ NSC_DecapsulateKey(CK_SESSION_HANDLE hSe
                         decapsulationKey->attrib.ulValueLen };
     SECItem ciphertext = { siBuffer, pCiphertext, ulCiphertextLen };
 
     /* The length of secretBuf can be increased if we ever support other KEMs
      * by changing the define at the top of this file */
     uint8_t secretBuf[MAX_KEM_SHARED_SECRET_BYTES] = { 0 };
     SECItem secret = { siBuffer, secretBuf, sizeof secretBuf };
     key->isFIPS = sftk_operationIsFIPS(slot, pMechanism, CKA_DECAPSULATE,
-                                       key, 0);
+                                       decapsulationKeyObject, 0);
     key->source = SFTK_SOURCE_KEA;
 
     switch (pMechanism->mechanism) {
 #ifndef NSS_DISABLE_KYBER
         case CKM_NSS_KYBER:
 #endif
         case CKM_NSS_ML_KEM:
         case CKM_ML_KEM:
diff --git a/lib/softoken/pkcs11c.c b/lib/softoken/pkcs11c.c
--- a/lib/softoken/pkcs11c.c
+++ b/lib/softoken/pkcs11c.c
@@ -5236,17 +5236,17 @@ sftk_compareKeysEqual(CK_SESSION_HANDLE 
     SFTKObject *key1obj = NULL;
     SFTKObject *key2obj = NULL;
     SFTKAttribute *att1 = NULL;
     SFTKAttribute *att2 = NULL;
 
     /* fetch the pkcs11 objects from the handles */
     session = sftk_SessionFromHandle(hSession);
     if (session == NULL) {
-        return CKR_SESSION_HANDLE_INVALID;
+        return PR_FALSE;
     }
 
     key1obj = sftk_ObjectFromHandle(key1, session);
     key2obj = sftk_ObjectFromHandle(key2, session);
     sftk_FreeSession(session);
     if ((key1obj == NULL) || (key2obj == NULL)) {
         goto  loser;
     }
@@ -5268,17 +5268,17 @@ sftk_compareKeysEqual(CK_SESSION_HANDLE 
         goto loser;
     }
     result = PR_TRUE;
 loser:
     if (key1obj) {
         sftk_FreeObject(key1obj);
     }
     if (key2obj) {
-        sftk_FreeObject(key1obj);
+        sftk_FreeObject(key2obj);
     }
     if (att1) {
         sftk_FreeAttribute(att1);
     }
     if (att2) {
         sftk_FreeAttribute(att2);
     }
     return result;
@@ -5758,24 +5758,24 @@ sftk_PairwiseConsistencyCheck(CK_SESSION
             SECITEM_ZfreeItem(&prime, PR_FALSE);
         }
         /* clean up before we return */
         sftk_FreeAttribute(pubAttribute);
         if (crv != CKR_OK) {
             return crv;
         }
     }
-    isKEM = sftk_isTrue(privateKey, CKA_ENCAPSULATE);
+    isKEM = sftk_isTrue(privateKey, CKA_DECAPSULATE);
     if (isKEM) {
         unsigned char *cipher_text = NULL;
         CK_ULONG cipher_text_length = 0;
         CK_OBJECT_HANDLE key1 =  CK_INVALID_HANDLE;
         CK_OBJECT_HANDLE key2 =  CK_INVALID_HANDLE;
         CK_KEY_TYPE genType = CKO_SECRET_KEY;
-        CK_ATTRIBUTE template = { CKA_KEY_TYPE, NULL, 0 };
+        CK_ATTRIBUTE template = { CKA_CLASS, NULL, 0 };
 
         template.pValue = &genType;
         template.ulValueLen = sizeof(genType);
         crv = CKR_OK;
         switch (keyType) {
             case CKK_ML_KEM:
                 cipher_text_length = MAX_ML_KEM_CIPHER_LENGTH;
                 mech.mechanism = CKM_ML_KEM;
@@ -5828,17 +5828,17 @@ kem_done:
         PORT_Free(cipher_text);
         if (key1 != CK_INVALID_HANDLE) {
             NSC_DestroyObject(hSession, key1);
         }
         if (key2 != CK_INVALID_HANDLE) {
             NSC_DestroyObject(hSession, key2);
         }
         if (crv != CKR_OK) {
-            return CKR_DEVICE_ERROR;
+            return crv;
         }
     }
 
     return CKR_OK;
 }
 
 /* NSC_GenerateKeyPair generates a public-key/private-key pair,
  * creating new key objects. */
@@ -7905,16 +7905,17 @@ sftk_DeriveEncrypt(SFTKCipher encrypt, v
         return crv;
     }
 
     crv = sftk_forceAttribute(key, CKA_VALUE, tmpdata, keySize);
     PORT_Memset(tmpdata, 0, sizeof tmpdata);
     return crv;
 }
 
+
 CK_RV
 sftk_HKDF(CK_HKDF_PARAMS_PTR params, CK_SESSION_HANDLE hSession,
           SFTKObject *sourceKey, const unsigned char *sourceKeyBytes,
           int sourceKeyLen, SFTKObject *key, unsigned char *outKeyBytes,
           int keySize, PRBool canBeData, PRBool isFIPS)
 {
     SFTKSession *session;
     SFTKAttribute *saltKey_att = NULL;
@@ -7958,16 +7959,21 @@ sftk_HKDF(CK_HKDF_PARAMS_PTR params, CK_
     }
 
     /* sourceKey is NULL if we are called from the POST, skip the
      * sensitiveCheck */
     if (sourceKey != NULL) {
         crv = sftk_DeriveSensitiveCheck(sourceKey, key, canBeData);
         if (crv != CKR_OK)
             return crv;
+         /* if the source key is data, clear the FIPS flag
+          * and only get the FIPS state from the salt */
+         if (sourceKey->objclass == CKO_DATA) {
+             key->isFIPS = PR_FALSE;
+         }
     }
 
     /* HKDF-Extract(salt, base key value) */
     if (params->bExtract) {
         CK_BYTE *salt;
         CK_ULONG saltLen;
         HMACContext *hmac;
         unsigned int bufLen;
@@ -7988,16 +7994,17 @@ sftk_HKDF(CK_HKDF_PARAMS_PTR params, CK_
                 break;
             case CKF_HKDF_SALT_KEY:
                 /* lookup key */
                 session = sftk_SessionFromHandle(hSession);
                 if (session == NULL) {
                     return CKR_SESSION_HANDLE_INVALID;
                 }
 
+
                 saltKey = sftk_ObjectFromHandle(params->hSaltKey, session);
                 sftk_FreeSession(session);
                 if (saltKey == NULL) {
                     return CKR_KEY_HANDLE_INVALID;
                 }
                 /* if the base key is not fips, but the salt key is, the
                  * resulting key can be fips */
                 if (isFIPS && (key->isFIPS == 0) && (saltKey->isFIPS == 1)) {
@@ -9190,17 +9197,23 @@ NSC_DeriveKey(CK_SESSION_HANDLE hSession
 
             PORT_Memcpy(buf, att->attrib.pValue, att->attrib.ulValueLen);
             PORT_Memcpy(buf + att->attrib.ulValueLen,
                         att2->attrib.pValue, att2->attrib.ulValueLen);
 
             crv = sftk_forceAttribute(key, CKA_VALUE, buf, keySize);
             PORT_ZFree(buf, tmpKeySize);
             /* preserve the source of the original base key */
-            /* key->source = sourceKey->source; */
+            key->source = sourceKey->source;
+
+            /* make sure this is fully fips approved, and mark it
+             * unapproved if not */
+            if (key->isFIPS) {
+                key->isFIPS = paramKey->isFIPS;
+            }
             sftk_FreeAttribute(att2);
             sftk_FreeObject(paramKey);
             break;
         }
 
         case CKM_CONCATENATE_BASE_AND_DATA:
             crv = sftk_DeriveSensitiveCheck(sourceKey, key, PR_FALSE);
             if (crv != CKR_OK)
diff --git a/lib/softoken/pkcs11u.c b/lib/softoken/pkcs11u.c
--- a/lib/softoken/pkcs11u.c
+++ b/lib/softoken/pkcs11u.c
@@ -2285,16 +2285,22 @@ sftk_AttributeToFlags(CK_ATTRIBUTE_TYPE 
             flags = CKF_VERIFY;
             break;
         case CKA_VERIFY_RECOVER:
             flags = CKF_VERIFY_RECOVER;
             break;
         case CKA_DERIVE:
             flags = CKF_DERIVE;
             break;
+        case CKA_ENCAPSULATE:
+            flags = CKF_ENCAPSULATE;
+            break;
+        case CKA_DECAPSULATE:
+            flags = CKF_DECAPSULATE;
+            break;
         /* fake attribute to select digesting */
         case CKA_DIGEST:
             flags = CKF_DIGEST;
             break;
         /* fake attribute to select key gen */
          case CKA_NSS_GENERATE:
             flags = CKF_GENERATE;
             break;
